src/main/kotlin/io/sc3/peripherals/posters/PosterItem.kt

```diff
@@ -0,0 +1,217 @@
+      val mapState = getOrCreatePosterState(stack, world)
+      if (mapState != null) {
+        if (entity is PlayerEntity) {
+          mapState.update(entity)
+        }
+      }
+    }
+  }
+
+  override fun getName(stack: ItemStack): Text = printData(stack)?.labelText ?: super.getName(stack)
+
+  override fun appendTooltip(stack: ItemStack, world: World?, tooltip: MutableList<Text>, context: TooltipContext) {
+    val id = getPosterId(stack)
+
+    val data = printData(stack) ?: return
+    data.tooltip?.let { tooltip.add(Text.literal(it)) }
+
+    if (context.isAdvanced) {
+      if (id != null) {
+        tooltip.add(Text.translatable("sc-peripherals.poster.id", id.take(8)).formatted(Formatting.GRAY))
+      } else {
+        tooltip.add(Text.translatable("sc-peripherals.poster.unknown").formatted(Formatting.GRAY))
+      }
+    }
+  }
+
+  data class PosterTooltipComponent(val stack: ItemStack, val world: World?) : TooltipComponent {
+    override fun drawItems(
+      textRenderer: TextRenderer,
+      tooltipX: Int,
+      tooltipY: Int,
+      matrices: MatrixStack,
+      itemRenderer: ItemRenderer,
+      something: Int
+    ) {
+      val posterId = getPosterId(stack) ?: return
+      val posterState = getPosterState(posterId, world) ?: return
+      RenderSystem.setShader(GameRenderer::getPositionTexProgram)
+      RenderSystem.setShaderColor(1.0f, 1.0f, 1.0f, 1.0f)
+      RenderSystem.setShaderTexture(0, POSTER_BACKGROUND_RES)
+      val pad = 7
+      val size = 135 + pad
+      val scale = 0.5f
+      matrices.push()
+      matrices.translate(tooltipX + 3.0, tooltipY + 3.0, 500.0)
+      matrices.scale(scale, scale, 1f)
+      RenderSystem.enableBlend()
+      DrawableHelper.drawTexture(matrices, -pad, -pad, 0f, 0f, size, size, size, size)
+      val buffer: BufferBuilder = Tessellator.getInstance().buffer
+      val immediateBuffer = VertexConsumerProvider.immediate(buffer)
+      PosterRenderer.draw(matrices, immediateBuffer, posterId, posterState, 240)
+      immediateBuffer.draw()
+      matrices.pop()
+    }
+
+    override fun getHeight(): Int {
+      val posterId = getPosterId(stack)
+      val posterState = getPosterState(posterId, world)
+      return if (posterState != null) 75 else 0
+    }
+
+    override fun getWidth(textRenderer: TextRenderer) = 72
+  }
+
+  companion object {
+    const val POSTER_KEY = "poster"
+
+    fun getPosterName(posterId: String) = "posters/${posterId.take(2)}/$posterId"
+
+    fun create(world: World, data: PosterPrintData): ItemStack {
+      val itemStack = ItemStack(ModItems.poster)
+      if (data.posterId != null) {
+        setPosterId(itemStack, data.posterId!!, data)
+      } else {
+        createPosterState(itemStack, world, data)
+      }
+
+      return itemStack
+    }
+
+    private fun setPosterId(stack: ItemStack, id: String, data: PosterPrintData) {
+      stack.orCreateNbt.putString(POSTER_KEY, id)
+      stack.nbt!!.copyFrom(data.toItemNbt())
```

Comment:
I'm at war with null assertions

```suggestion
      with (stack.orCreateNbt) {
        putString(POSTER_KEY, id)
        copyFrom(data.toItemNbt())
      }
```